### PR TITLE
Fix CLuaFunctionParserBase lua_State* index check

### PR DESCRIPTION
**Pull request by qaisjp**

### CODE REVIEW

**Comment by ghost** [comment]
merge?

**Comment by qaisjp** [comment]
sbx is working on an alt solution

File: Shared/mods/deathmatch/logic/lua/CLuaFunctionParser.h [hunk 1]
```diff
@@ -212,7 +214,11 @@ struct CLuaFunctionParserBase
 
         // lua_State* can be taken as first argument of any function
         if constexpr (std::is_same_v<T, lua_State*>)
-            return index == 0;
+        {
+            bool constraint = index == 1;
+            dassert(constraint && "`lua_State*` must only be placed in the first parameter of Lua definitions");
+            return constraint;
```
Comment (qaisjp) [inline Shared/mods/deathmatch/logic/lua/CLuaFunctionParser.h]:
```suggestion
            return index == 1;
```

File: Shared/mods/deathmatch/logic/lua/CLuaFunctionParser.h [hunk 2]
```diff
@@ -212,7 +214,9 @@ struct CLuaFunctionParserBase
 
         // lua_State* can be taken as first argument of any function
         if constexpr (std::is_same_v<T, lua_State*>)
-            return index == 0;
+        {
```
Comment (qaisjp) [inline Shared/mods/deathmatch/logic/lua/CLuaFunctionParser.h]:
```suggestion
```

File: Shared/mods/deathmatch/logic/lua/CLuaFunctionParser.h [hunk 3]
```diff
@@ -212,7 +214,9 @@ struct CLuaFunctionParserBase
...
+            return index == 1;
+        }
```
Comment (qaisjp) [inline Shared/mods/deathmatch/logic/lua/CLuaFunctionParser.h]:
```suggestion
```

**Review by ghost — APPROVED** [review]

**Pull request merged**